plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -79,97 +82,124 @@ Then, in the _body_, we track the term and type application (`Apply` or `TyInst`
-e.g. consider the term @let id = \x -> x in id@: the variable @id@ has syntactic
-arity @[]@, but does in fact need an argument before it does any work.
--}
-type Arity = [TermOrType]
-
--- | Is the next argument a term or a type?
-data TermOrType =
-    MkTerm | MkType
-    deriving stock (Eq, Show)
-
-instance Pretty TermOrType where
-  pretty = viaShow
-
--- | Computes the 'Arity' of a term.
+-- | Computes the 'Utils.Arity' of a term. Also returns the function body, for checking whether
+-- it's `Utils.acceptable`.
 computeArity ::
-    Term tyname name uni fun a
-    -> Arity
+    Term tyname name uni fun ann
+    -> (Arity, Term tyname name uni fun ann)
 computeArity = \case
-    LamAbs _ _ _ body -> MkTerm : computeArity body
-    TyAbs _ _ _ body  -> MkType : computeArity body
+    LamAbs _ _ _ body -> (first ((:) MkTerm) (computeArity body))
+    TyAbs _ _ _ body  -> (first ((:) MkType) (computeArity body))
     -- Whenever we encounter a body that is not a lambda or type abstraction, we are done counting
-    _                 -> []
+    tm                -> ([],tm)
+
+-- | A term or type argument.
+data Args tyname name uni fun ann =
+  MkTermArg (Term tyname name uni fun ann)
+  | MkTypeArg (Type tyname uni ann)
+
+-- | A list of type or term argument(s) that are being applied.
+type ArgOrder tyname name uni fun ann = [Args tyname name uni fun ann]
+
+-- | A pair of argument and the annotation of the term being applied to,
+-- so a term that was traversed can be built back with `mkApps`.
+type ArgOrderWithAnn tyname name uni fun ann =
+  [(Args tyname name uni fun ann, ann)]
+
+-- | Takes a term or type application expression and returns the function
+-- being applied and the arguments to which it is applied
+collectArgs :: Term tyname name uni fun ann
+  -> (Term tyname name uni fun ann, ArgOrderWithAnn tyname name uni fun ann)
+collectArgs expr
+  = go expr []
+  where
+    go (Apply ann f a) as      = go f ((MkTermArg a, ann):as)
+    go (TyInst ann f tyArg) as = go f ((MkTypeArg tyArg, ann):as)
+    go e as                    = (e, as)
+
+-- | Apply a list of term and type arguments to a function in potentially a nested fashion.
+mkApps :: Term tyname name uni fun ann
+  -> ArgOrderWithAnn tyname name uni fun ann
+  -> Term tyname name uni fun ann
+mkApps f ((MkTermArg tmArg, ann) : args) = mkApps (Apply ann f tmArg) args
+mkApps f ((MkTypeArg tyArg, ann) : args) = mkApps (TyInst ann f tyArg) args
+mkApps f []                              = f
+
+-- | Given the arity of a function, and the list of arguments applied to it, return whether it is
+-- fully applied or not.
+isFullyApplied :: Arity -> ArgOrder tyname name uni fun ann -> Bool
+isFullyApplied [] (_argsOrder:_as) = True -- over-application
+isFullyApplied (_arity:_) [] = False -- under-application
+isFullyApplied [] [] = True
+isFullyApplied lamOrder argsOrder =
+  -- start comparing from the end because there may be over-application
```

Comment:
I'm also confused. I don't see why this doesn't just compare the two things at the heads of the list and then recurse.

Reply:
Do we have any test for over/under-application? Not sure that we do.

Reply:
I can reverse the lists and do the regular thing. I thought if the arg list is long this may be wasteful but as Michael pointed out this is not fast either.

Reply:
> As for collectArgs, it returns an ArgOrder with its head being the outer most application

I don't think this is true. Given `fun a b c d e f`, your `collectArgs` returns `(fun, [a, b, c, d, e, f])`, as far as I can tell. So you should be comparing the heads of both lists. If `fun` is the function you wrote above, then `a` should be compared against `MkType`, and `f` should be ignored.

And if I was wrong and `collectArgs` actually returns `(fun, [f, e, d, c, b, a])`, then it is still incorrect because you should compare the last element of one list against the first element of the other list.

Reply:
OK I changed to comparing the heads and it didn't change the test results. I don't know why it *was* working - maybe the tests aren't covering it properly...